[core] fix(Overlay): use focus traps to fix enforceFocus

michael-yx-wu opened the pull request:
#### Fixes https://github.com/palantir/blueprint/issues/3445

#### Checklist

- [x] Includes tests
- [x] Update documentation (n/a)

<!-- DO NOT enable CircleCI for your fork. Our build will run when you open this PR. -->

#### Changes proposed in this pull request:

When enforceFocus is enabled, surround Overlay content with dummy div
elements with focus event listeners that keep focus inside the Overlay.

#### Reviewers should focus on:

I want to flag that I've removed the `tabindex` attribute from the backdrop to prevent it from being keyboard focusable.

This change also fixes an issue where multiple open overlays will compete with each for focus. As an example, go to https://blueprintjs.com/docs/#core/components/overlay and open the overlay and then hit shift+s to open the omnibar search (also powered by Overlay). The search input will not get autofocused. Repeat the same workflow on https://46971-71939872-gh.circle-artifacts.com/0/packages/docs-app/dist/index.html#core/components/overlay and note that the omnibar search does get focus correctly.

#### Screenshot

##### Before

When using a portal, the portal is the last element on the page so tabbing will eventually cause focus to leave the document and none of the existing event handlers run before that happens. Note that shift+tabbing is not closing the overlay, the gif is just looping.

![2021-09-09 19 59 13](https://user-images.githubusercontent.com/3681045/132777763-9172c91e-f09b-449a-be2e-b73640513f0b.gif)

##### After

Tab/shift+tab will never allow focus to leave the overlay if `enforceFocus`.

![2021-09-09 19 58 14](https://user-images.githubusercontent.com/3681045/132777789-519292e2-b57b-48b1-8fe1-271101143d2e.gif)

michael-yx-wu committed:
Use focus traps to fix Overlay enforceFocus

When enforceFocus is enabled, surround Overlay content with dummy div
elements with focus event listeners that keep focus inside the Overlay.

michael-yx-wu committed:
Make it work without portals

michael-yx-wu committed:
Fix tests

michael-yx-wu commented on packages/core/src/components/overlay/overlay.tsx:
```diff
@@ -475,25 +553,14 @@ export class Overlay extends AbstractPureComponent2<OverlayProps, IOverlayState>
-        }
-    };
-
-    private handleDocumentFocus = (e: FocusEvent) => {
-        // get the actually target even if we are in an open mode Shadow DOM
-        const eventTarget = e.composed ? e.composedPath()[0] : e.target;
-        if (
-            this.props.enforceFocus &&
-            this.containerElement != null &&
-            eventTarget instanceof Node &&
-            !this.containerElement.contains(eventTarget as HTMLElement)
-        ) {
-            // prevent default focus behavior (sometimes auto-scrolls the page)
-            e.preventDefault();
-            e.stopImmediatePropagation();
-            this.bringFocusInsideOverlay();
```
when `enforceFocus=true`, focus will start on the overlay and will only leave if another overlay with autoFocus or enforceFocus is opened. because we take care to focus the last opened overlay when closing the current overlay and we call `bringFocusInsideOverlay` in the mousedown event handler, we don't need this focus event handler anymore

michael-yx-wu commented on packages/core/src/components/overlay/overlay.tsx:
```diff
@@ -415,7 +501,7 @@ export class Overlay extends AbstractPureComponent2<OverlayProps, IOverlayState>
                 const lastOpenedOverlay = Overlay.getLastOpened();
                 if (lastOpenedOverlay.props.enforceFocus) {
-                    document.addEventListener("focus", lastOpenedOverlay.handleDocumentFocus, /* useCapture */ true);
+                    lastOpenedOverlay.bringFocusInsideOverlay();
```
this handler was a roundabout way to bring focus back to the last opened overlay. we can just call `bringFocusInsideOverlay` directly.

michael-yx-wu commented on packages/core/test/multistep-dialog/multistepDialogTests.tsx:
```diff
@@ -15,18 +15,47 @@
+     * Mount the `content` into `testsContainerElement` and assign to local `wrapper` variable.
+     * Use this method in this suite instead of Enzyme's `mount` method.
+     */
+    function mountWrapper(content: JSX.Element) {
+        wrapper = mount(content, { attachTo: testsContainerElement });
+        isMounted = true;
+        return wrapper;
+    }
```
these tests were causing Overlay tests to fail. see https://app.circleci.com/pipelines/github/palantir/blueprint/1539/workflows/f959324a-857b-4131-a89e-f3c315416103/jobs/46941: 

> \<Overlay\> Focus management does not focus overlay when closed - <Overlay> Focus management
> `Error: Uncaught AssertionError: expected <button></button> to equal <div class="bp3-dialog-container bp3-overlay-content bp3-overlay-appear-done bp3-overlay-enter-done" tabindex="0"><div class="bp3-dialog" role="dialog" style="min-width: 800px; padding-bottom: 0px;"><div class="bp3-multistep-dialog-panels"><div class="bp3-multistep-dialog-left-panel"></div></div></div></div>`

Note the **`bp3-multistep-dialog-left-panel`** class. The error suggests that something is causing these tests to affect the overlay tests but I'm not sure why...

Making this suite use a mountWrapper like the overlayTests seems to fix the issue.

michael-yx-wu replied:
on closer inspection, I suspect the reason is due to some of these tests previously mounting but _not_ unmounting. For example: https://github.com/palantir/blueprint/blob/88a158b38a96350109650a7fe821a5091a25b3ed/packages/core/test/multistep-dialog/multistepDialogTests.tsx#L169-L180

Since this suite is alphabetically before overlayTests and because we've now fixed the `enforceFocus` behavior (which is enabled for both Dialog and MultistepDialog), these unmounted overlays are keeping focus. I'll revert this change and just unmount properly for a smaller diff, but lmk if you actually prefer this wrapper instead.

adidahiya replied:
unmounting properly sounds like a better change to do in this PR :+1:

michael-yx-wu committed:
Fix tests by unmounting properly

michael-yx-wu commented on packages/core/test/overlay/overlayTests.tsx:
```diff
@@ -60,6 +60,10 @@ describe("<Overlay>", () => {
         }
     });
 
+    after(() => {
+        document.documentElement.removeChild(testsContainerElement);
+    });
```
not really important, but seems good to clean up?

michael-yx-wu commented on packages/core/src/components/overlay/overlay.tsx:
```diff
@@ -319,11 +335,11 @@ export class Overlay extends AbstractPureComponent2<OverlayProps, IOverlayState>
             if (isFocusOutsideModal) {
                 // element marked autofocus has higher priority than the other clowns
                 const autofocusElement = this.containerElement.querySelector("[autofocus]") as HTMLElement;
```
I noticed that `InputGroup`'s `autoFocus` prop is passed to the `input` element as is. React doesn't actually render the autofocus attribute to the DOM (see https://github.com/facebook/react/issues/11851#issuecomment-351672131), so this querySelector is unlikely to work unless the consumer explicitly sets the DOM `autofocus` (no caps) attribute instead. Not sure if we want to do anything about this though, since React will warn if you set the DOM attribute directly

adidahiya replied:
we should do something about this in a FLUP. not a high priority though

michael-yx-wu replied:
Filed https://github.com/palantir/blueprint/issues/4908

adidahiya commented on packages/core/src/components/overlay/overlay.tsx:
```diff
@@ -229,10 +229,22 @@ export class Overlay extends AbstractPureComponent2<OverlayProps, IOverlayState>
     // an HTMLElement that contains the backdrop and any children, to query for focus target
     public containerElement: HTMLElement | null = null;
 
+    private startFocusTrapElement: HTMLDivElement | null = null;
```
would be nice to leave a brief `// comment` here about this member's purpose. same for `endFocusTrapElement`.

adidahiya commented on packages/core/src/components/overlay/overlay.tsx:
```diff
@@ -404,9 +419,80 @@ export class Overlay extends AbstractPureComponent2<OverlayProps, IOverlayState>
+        }
+    };
+
+    private keyboardFocusableElements() {
```
nit:

```suggestion
    private getKeyboardFocusableElements() {
```

adidahiya commented on packages/core/src/components/overlay/overlay.tsx:
```diff
@@ -404,9 +419,80 @@ export class Overlay extends AbstractPureComponent2<OverlayProps, IOverlayState>
+                ? Array.from(
+                      // Order may not be correct if children elements use tabindex values > 0
+                      this.containerElement.querySelectorAll(
+                          [
```
I'm curious how you decided on this list? Is there some documentation we can link to? It would be nice to leave some more comments about our reasoning for this code

michael-yx-wu replied:
I cobbled this together from jquery-ui's [`focusable` selector](https://github.com/jquery/jquery-ui/blob/70dae67b73dfea9126f126f516fe8286f1e73417/ui/focusable.js#L45) and this [SO answer](https://stackoverflow.com/a/1600194/2213364). I will add a comment

adidahiya reviewed (commented):
overall looks pretty good!

michael-yx-wu committed:
Add focus trap element comments

michael-yx-wu committed:
Ignore tabindex="-1" elements

michael-yx-wu committed:
Fallback to focus on startFocusTrapElement

This allows bringFocusInsideOverlay to work even when there are no
keyboard-focusable elements.

michael-yx-wu commented:
I found a bug where this doesn't properly enforce focus if focus is coming back to the document _from_ another part of the browser (e.g. the URL or bookmarks bar). I will restore the focus event handler logic, but I will listen instead for `focusin` instead of `focus` since the latter isn't supposed to bubble

update: as suggested, I'll address this in a flup

michael-yx-wu committed:
Fix selector

adidahiya reviewed (approved):
nice :+1:

ro-savage commented:
@michael-yx-wu - The appears to be a bug or unintentional behaviour with this change on dialogs with scrolls.

See: #4920
